scripts/generate-unicode-id-parts.js

```diff
@@ -0,0 +1,158 @@
+      ']'
+    );
+  }
+}
+
+class TrailSurrogateRange {
+  /** @param {number} trailSurrogate */
+  constructor(trailSurrogate) {
+    this.firstTrailSurrogate = trailSurrogate;
+    this.lastTrailSurrogate = trailSurrogate;
+  }
+  toString() {
+    let text = '\\u' + toHex4Digits(this.firstTrailSurrogate);
+    if (this.lastTrailSurrogate !== this.firstTrailSurrogate) {
+      text += '-\\u' + toHex4Digits(this.lastTrailSurrogate);
+    }
+    return text;
+  }
+}
+
+class Writer {
+  constructor() {
+    /** @type {number} */
+    this.lastCodePoint = -1;
+    /** @type {NonSurrogateRange[]} */
+    this.nonSurrogateRanges = [];
+    /** @type {LeadSurrogateRange[]} */
+    this.surrogateRanges = [];
+    /** @type {NonSurrogateRange} */
+    this.nonSurrogateRange;
+    /** @type {LeadSurrogateRange} */
+    this.leadSurrogateRange;
+    /** @type {TrailSurrogateRange} */
+    this.trailSurrogateRange;
+  }
+
+  /** @param {number} codePoint */
+  push(codePoint) {
+    if (codePoint <= this.lastCodePoint)
+      throw new Error('Code points must be added in order.');
+    this.lastCodePoint = codePoint;
+
+    if (codePoint < MAX_UNICODE_NON_SURROGATE) {
+      if (
+        this.nonSurrogateRange &&
+        this.nonSurrogateRange.lastCodePoint === codePoint - 1
+      ) {
+        this.nonSurrogateRange.lastCodePoint = codePoint;
+        return;
+      }
+      this.nonSurrogateRange = new NonSurrogateRange(codePoint);
+      this.nonSurrogateRanges.push(this.nonSurrogateRange);
+    } else {
+      const leadSurrogate = Math.floor((codePoint - 0x10000) / 0x400) + 0xd800;
+      const trailSurrogate = ((codePoint - 0x10000) % 0x400) + 0xdc00;
+      if (
+        !this.leadSurrogateRange ||
+        this.leadSurrogateRange.leadSurrogate !== leadSurrogate
+      ) {
+        this.trailSurrogateRange = undefined;
+        this.leadSurrogateRange = new LeadSurrogateRange(leadSurrogate);
+        this.surrogateRanges.push(this.leadSurrogateRange);
+      }
+
+      if (
+        this.trailSurrogateRange &&
+        this.trailSurrogateRange.lastTrailSurrogate === trailSurrogate - 1
+      ) {
+        this.trailSurrogateRange.lastTrailSurrogate = trailSurrogate;
+        return;
+      }
+
+      this.trailSurrogateRange = new TrailSurrogateRange(trailSurrogate);
+      this.leadSurrogateRange.ranges.push(this.trailSurrogateRange);
+    }
+  }
+
+  toString() {
+    let first = this.nonSurrogateRanges.join('');
+    let second = this.surrogateRanges.join('|');
+    return first && second
+      ? `([${first}]|${second})`
+      : first
+      ? `[${first}]`
+      : second
+      ? `(${second})`
+      : '';
+  }
+}
+
+const MAX_UNICODE_NON_SURROGATE = 0xffff;
+const MAX_UNICODE_CODEPOINT = 0x10ffff;
+const isStart = c => /\p{ID_Start}/u.test(c);
+const isContinue = c => /\p{ID_Continue}/u.test(c);
+
+let idStartWriter = new Writer();
+let idContinueWriter = new Writer();
+
+for (let cp = 0; cp < MAX_UNICODE_CODEPOINT; cp++) {
```

Comment:
This should have been `<=`

Reply:
Not sure that it matters that much, 0x10ffff is in the SPUA and likely won't ever be part of either unicode property.
```
> /[\p{ID_Start}\p{ID_Continue}]/u.test(String.fromCodePoint(0x10ffff))
false
```

Reply:
I know, but if it was intentional, why stop at U+10FFFE?

```js
/[\p{ID_Start}\p{ID_Continue}]/u.test('\u{10fffe}');
// --> false
```

Reply:
Not intentional, and I've pushed a fix. I was just commenting that it didn't affect the output.